File: funcs.go
package spacemonger

import (
	"context"
	"fmt"
	"log"
	"regexp"
	"strconv"
	"time"

	"spacemonger/spacetraders"
)

type FlightPlan struct {
	Id                     string
	ShipId                 string
	FuelConsumed           int
	FuelRemaining          int
	TimeRemainingInSeconds int
	CreatedAt              time.Time
	ArrivesAt              time.Time
	TerminatedAt           time.Time
	Destination            string
	Origin                 string
	Distance               int
}

func CreateFlightPlan(ctx context.Context, client spacetraders.AuthorizedClient, conn DbConn, s *Ship, destination string) (FlightPlan, error) {
	flightPlanResp, err := client.CreateFlightPlan(ctx, s.Id, destination)
	if err != nil {
		return FlightPlan{}, err
	}

	s.Location = ""
	var newCargo []Cargo
	for _, c := range s.Cargo {
		if c.Good == GoodFuel {
			c.Quantity -= flightPlanResp.FlightPlan.FuelConsumed
		}

		newCargo = append(newCargo, c)
	}

	s.Cargo = newCargo

	err = SaveFlightPlan(ctx, conn, s.UserId, DbFlightPlan{
		Id:                     flightPlanResp.FlightPlan.Id,
		UserId:                 s.UserId,
		ShipId:                 s.Id,
		Origin:                 flightPlanResp.FlightPlan.Departure,
		Destination:            flightPlanResp.FlightPlan.Destination,
		Distance:               flightPlanResp.FlightPlan.Distance,
		FuelConsumed:           flightPlanResp.FlightPlan.FuelConsumed,
		FuelRemaining:          flightPlanResp.FlightPlan.FuelRemaining,
		TimeRemainingInSeconds: flightPlanResp.FlightPlan.TimeRemainingInSeconds,
		ArrivesAt:              flightPlanResp.FlightPlan.ArrivesAt,
		CreatedAt:              flightPlanResp.FlightPlan.CreatedAt,
	})
	if err != nil {
		return FlightPlan{}, err
	}

	return FlightPlan{
		Id:                     flightPlanResp.FlightPlan.Id,
		ShipId:                 s.Id,
		FuelConsumed:           flightPlanResp.FlightPlan.FuelConsumed,
		FuelRemaining:          flightPlanResp.FlightPlan.FuelRemaining,
		TimeRemainingInSeconds: flightPlanResp.FlightPlan.TimeRemainingInSeconds,
		CreatedAt:              flightPlanResp.FlightPlan.CreatedAt,
		ArrivesAt:              flightPlanResp.FlightPlan.ArrivesAt,
		TerminatedAt:           flightPlanResp.FlightPlan.TerminatedAt,
		Origin:                 flightPlanResp.FlightPlan.Departure,
		Destination:            flightPlanResp.FlightPlan.Destination,
		Distance:               flightPlanResp.FlightPlan.Distance,
	}, nil
}

func CreatePurchaseOrder(ctx context.Context, client spacetraders.AuthorizedClient, conn DbConn, s *Ship, good string, quantity int) (int, error) {
	if quantity > 0 {
		resp, err := client.CreatePurchaseOrder(ctx, s.Id, good, quantity)
		if err != nil {
			return 0, fmt.Errorf("unable to create purchase order for good \"%s\" quantity \"%d\": %w", good, quantity, err)
		}

		var newCargo []Cargo
		for _, c := range resp.Ship.Cargo {
			newCargo = append(newCargo, Cargo(c))
		}

		s.Cargo = newCargo
		s.SpaceAvailable = resp.Ship.SpaceAvailable

		err = SaveTransaction(ctx, conn, DbTransaction{
			UserId:       s.UserId,
			ShipId:       s.Id,
			Type:         "purchase",
			Good:         good,
			PricePerUnit: resp.Order.PricePerUnit,
			Quantity:     resp.Order.Quantity,
			Total:        resp.Order.Total,
			Location:     resp.Ship.Location,
		})

		if err != nil {
			return 0, fmt.Errorf("unable to save purchase order to db: %w", err)
		}

		return resp.Credits, nil
	} else {
		return 0, fmt.Errorf("refusing to attempt to create purchase order with 0 quantity")
	}
}

func CreateSellOrder(ctx context.Context, client spacetraders.AuthorizedClient, conn DbConn, s *Ship, good string, quantity int) (int, error) {
	if quantity > 0 {
		resp, err := client.CreateSellOrder(ctx, s.Id, good, quantity)
		if err != nil {
			return 0, fmt.Errorf("unable to create sell order for good \"%s\" quantity \"%d\": %w", good, quantity, err)
		}

		var newCargo []Cargo
		for _, c := range resp.Ship.Cargo {
			newCargo = append(newCargo, Cargo(c))
		}

		s.Cargo = newCargo
		s.SpaceAvailable = resp.Ship.SpaceAvailable

		err = SaveTransaction(ctx, conn, DbTransaction{
			UserId:       s.UserId,
			ShipId:       s.Id,
			Type:         "sell",
			Good:         good,
			PricePerUnit: resp.Order.PricePerUnit,
			Quantity:     resp.Order.Quantity,
			Total:        resp.Order.Total,
			Location:     resp.Ship.Location,
		})

		if err != nil {
			return 0, fmt.Errorf("unable to save transaction to db: %w", err)
		}

		return resp.Credits, nil
	} else {
		return 0, fmt.Errorf("refusing to attempt to create sell order with 0 quantity")
	}
}

func GetAdditionalFuelRequiredForTrip(ctx context.Context, client spacetraders.AuthorizedClient, conn DbConn, s Ship, destination string) (int, error) {
	currentFuel := 0
	for _, c := range s.Cargo {
		if c.Good == GoodFuel {
			currentFuel += c.Quantity
		}
	}

	log.Printf("%s:%s -- Ship currently has %d fuel\n", s.Username, s.Id, currentFuel)

	dbFuelRequired, err := GetFuelRequired(ctx, conn, s.Location, destination, s.Type)
	if err == nil {
		log.Printf("%s:%s -- Using fuel required from the db %d\n", s.Username, s.Id, dbFuelRequired)
		if dbFuelRequired > currentFuel {
			return dbFuelRequired - currentFuel, nil
		}

		return 0, nil
	}

	// At this point we weren't able to look up the fuel required so we should prepare our ship by selling all the fuel,
	// attempting to make a flight plan and return how much fuel is required to make the flight

	log.Printf("%s:%s -- Selling %d fuel in order to get fuel required from Api\n", s.Username, s.Id, currentFuel)

	if currentFuel > 0 {
		sellOrder, err := client.CreateSellOrder(ctx, s.Id, GoodFuel, currentFuel)
		if err != nil {
			return 0, err
		}

		s.Messages <- ShipMessage{
			Type:       ShipMessageUpdateCredits,
			ShipId:     s.Id,
			NewCredits: sellOrder.Credits,
		}
	}

	_, err = client.CreateFlightPlan(ctx, s.Id, destination)
	if err == nil {
		return 0, fmt.Errorf("create flight plan should have failed... ship is now in motion")
	}

	log.Printf("%s:%s -- Received error message from CreateFlightPlan (this is expected) %s\n", s.Username, s.Id, err.Error())

	re := regexp.MustCompile(`You require (\d+?) more FUEL`) // want to know what is in front of 'at'
	requiredFuelMatches := re.FindStringSubmatch(err.Error())

	log.Printf("%s:%s -- %s additional fuel is required", s.Username, s.Id, requiredFuelMatches[1])

	requiredFuel, err := strconv.Atoi(requiredFuelMatches[1])
	if err != nil {
		return 0, err
	}

	return requiredFuel, nil
}

// PurchaseFastestShip will attempt to purchase a new ship for the user. If no ship was able to be purchased then the
// original unmodified user will be returned along with the error.
func PurchaseFastestShip(ctx context.Context, u User, system string) (spacetraders.Ship, int, error) {
	availableShips, err := u.Client.GetShipsForSale(ctx)
	if err != nil {
		return spacetraders.Ship{}, 0, err
	}

	currentShips, err := u.Client.GetMyShips(ctx)
	if err != nil {
		return spacetraders.Ship{}, 0, err
	}

	dockedShipLocations := make(map[string]bool)
	for _, ship := range currentShips.Ships {
		if ship.Location != "" {
			dockedShipLocations[ship.Location] = true
		}
	}

	log.Printf("%s -- Docked ship locations are %v\n", u.Username, dockedShipLocations)
	log.Printf("%s -- User has %d ships\n", u.Username, len(u.Ships))
	log.Printf("%s -- Ships available for purchase %+v\n", u.Username, availableShips)

	if len(u.Ships) > 0 && len(dockedShipLocations) == 0 {
		log.Printf("%s -- No docked ships found. Unable to purchase new ship. Will retry later\n", u.Username)
		return spacetraders.Ship{}, 0, nil
	}

	fastestShipSpeed := 0
	fastestShipPrice := 0
	fastestShipLocation := ""
	fastestShipType := ""
	foundShip := false

	for _, availableShip := range availableShips.ShipsForSale {
		for _, purchaseLocation := range availableShip.PurchaseLocations {
			// users can only purchase ships at locations where they have a ship docked...
			// unless they currently don't have any ships
			if _, ok := dockedShipLocations[purchaseLocation.Location]; !ok && len(u.Ships) > 0 {
				continue
			}

			// TODO: Handle restricted goods better. Right now I just ignore any ships that are restricted
			//       to specific goods
			if len(availableShip.RestrictedGoods) > 0 {
				continue
			}

			if u.Credits < purchaseLocation.Price {
				continue
			}

			if availableShip.Speed < fastestShipSpeed {
				continue
			}

			if purchaseLocation.System != system {
				continue
			}

			foundShip = true
			fastestShipSpeed = availableShip.Speed
			fastestShipType = availableShip.Type
			fastestShipLocation = purchaseLocation.Location
			fastestShipPrice = purchaseLocation.Price
		}
	}

	if !foundShip {
		return spacetraders.Ship{}, 0, fmt.Errorf("%s -- unable to find a ship for the user to purchase", u.Username)
	}

	log.Printf("%s -- Buying ship %s for %d at location %s\n", u.Username, fastestShipType, fastestShipPrice, fastestShipLocation)
	s, err := u.Client.PurchaseShip(ctx, fastestShipLocation, fastestShipType)
	if err != nil {
		return spacetraders.Ship{}, 0, err
	}

	return s.Ship, s.Credits, nil
}

// PurchaseShip will attempt to purchase a new ship for the user. If no ship was able to be purchased then the
// original unmodified user will be returned along with the error.
func PurchaseShip(ctx context.Context, u User, system string, shipType string) (spacetraders.Ship, int, error) {
	availableShips, err := u.Client.GetShipsForSale(ctx)
	if err != nil {
		return spacetraders.Ship{}, 0, err
	}

	currentShips, err := u.Client.GetMyShips(ctx)
	if err != nil {
		return spacetraders.Ship{}, 0, err
	}

	dockedShipLocations := make(map[string]bool)
	for _, ship := range currentShips.Ships {
		if ship.Location != "" {
			dockedShipLocations[ship.Location] = true
		}
	}

	if len(u.Ships) > 0 && len(dockedShipLocations) == 0 {
		log.Printf("%s -- No docked ships found. Unable to purchase new ship. Will retry later\n", u.Username)
		return spacetraders.Ship{}, 0, nil
	}

	foundShip := false
	shipPrice := 0
	shipLocation := ""

	for _, availableShip := range availableShips.ShipsForSale {
		for _, purchaseLocation := range availableShip.PurchaseLocations {
			// users can only purchase ships at locations where they have a ship docked...
			// unless they currently don't have any ships
			if _, ok := dockedShipLocations[purchaseLocation.Location]; !ok && len(u.Ships) > 0 {
				continue
			}

			// TODO: Handle restricted goods better. Right now I just ignore any ships that are restricted
			//       to specific goods
			if len(availableShip.RestrictedGoods) > 0 {
				continue
			}

			if purchaseLocation.System != system {
				continue
			}

			if availableShip.Type != shipType {
				continue
			}

			foundShip = true
			shipPrice = purchaseLocation.Price
			shipLocation = purchaseLocation.Location
		}
	}

	if !foundShip {
		log.Printf("%s -- Docked ship locations are %v\n", u.Username, dockedShipLocations)
		log.Printf("%s -- User has %d ships\n", u.Username, len(u.Ships))
		log.Printf("%s -- Ships available for purchase %+v\n", u.Username, availableShips)

		return spacetraders.Ship{}, 0, fmt.Errorf("%s -- unable to find a ship for the user to purchase", u.Username)
	}

	log.Printf("%s -- Buying ship %s for %d at location %s\n", u.Username, shipType, shipPrice, shipLocation)
	s, err := u.Client.PurchaseShip(ctx, shipLocation, shipType)
	if err != nil {
		log.Printf("%s -- Error puchasing ship %v", u.Username, err)
		return spacetraders.Ship{}, 0, err
	}

	log.Printf("%s -- User purchased ship %+v\n", u.Username, s)

	return s.Ship, s.Credits, nil
}

type Route struct {
	PurchaseLocation          string
	PurchaseLocationType      string
	SellLocation              string
	Good                      string
	Distance                  float64
	PurchaseLocationQuantity  int
	SellLocationQuantity      int
	PurchasePricePerUnit      int
	SellPricePerUnit          int
	VolumePerUnit             int
	CostVolumeDistance        float64
	ProfitSpeedVolumeDistance float64
}

func GetBestTradingRoute(ctx context.Context, conn DbConn, s Ship) (Route, error) {
	log.Printf("Getting routes for ship from location \"%s\"\n", s.Location)
	dbRoutes, err := GetRoutesFromLocation(ctx, conn, s.Location)
	if err != nil {
		return Route{}, fmt.Errorf("unable to get routes from db: %w", err)
	}

	foundARoute := false
	bestRoute := Route{}
	for _, r := range dbRoutes {
		profit := float64(r.SellPricePerUnit - r.PurchasePricePerUnit)
		costVolumeDistance := profit / float64(r.VolumePerUnit) / r.Distance
		profitSpeedVolumeDistance := (profit * float64(s.Speed)) / (float64(r.VolumePerUnit) * r.Distance)

		route := Route{
			PurchaseLocation:          r.PurchaseLocation,
			PurchaseLocationType:      r.PurchaseLocationType,
			SellLocation:              r.SellLocation,
			Good:                      r.Good,
			Distance:                  r.Distance,
			PurchaseLocationQuantity:  r.PurchaseLocationQuantity,
			SellLocationQuantity:      r.SellLocationQuantity,
			PurchasePricePerUnit:      r.PurchasePricePerUnit,
			SellPricePerUnit:          r.SellPricePerUnit,
			VolumePerUnit:             r.VolumePerUnit,
			CostVolumeDistance:        costVolumeDistance,
			ProfitSpeedVolumeDistance: profitSpeedVolumeDistance,
		}

		if route.SellLocation == "OE-XV-91-2" {
			continue
		}

		if route.PurchaseLocationQuantity < 500 {
			continue
		}

		// We must allow trades that will cost us money so we don't get stuck at any location
		// if route.ProfitSpeedVolumeDistance <= 0.0 {
		// 	continue
		// }

		if !foundARoute || (route.ProfitSpeedVolumeDistance > bestRoute.ProfitSpeedVolumeDistance) {
			foundARoute = true
			bestRoute = route
		}
	}

	if !foundARoute {
		return Route{}, fmt.Errorf("unable to find any trade route from \"%s\"", s.Location)
	}

	return bestRoute, nil
}

File: daemon/src/funcs.rs
use anyhow::anyhow;
use spacetraders::client::{self, HttpClient, Client};
use spacetraders::errors::SpaceTradersClientError;
use sqlx::PgPool;
use spacetraders::responses;
use crate::db;
use spacetraders::shared::Good;
use crate::db::DbRoute;
use regex::Regex;

pub async fn is_api_in_maintenance_mode(http_client: HttpClient) -> bool {
    let game_status = client::get_game_status(http_client.clone()).await;

    if game_status.is_err() {
        let game_status_error = game_status.err().unwrap();

        return matches!(game_status_error, SpaceTradersClientError::ServiceUnavailable)
    }

    false
}

pub async fn create_flight_plan(client: Client, pg_pool: PgPool, user_id: &str, ship_id: &str, destination: &str) -> anyhow::Result<responses::FlightPlan> {
    let flight_plan = client.create_flight_plan(ship_id.to_string(), destination.to_string()).await?;

    db::persist_flight_plan(pg_pool, user_id, ship_id, &flight_plan).await?;

    Ok(flight_plan)
}

pub async fn create_purchase_order(client: Client, pg_pool: PgPool, user_id: &str, ship_id: &str, good: Good, quantity: i32) -> anyhow::Result<responses::PurchaseOrder> {
    if quantity > 0 {
        let purchase_order = client.create_purchase_order(ship_id.to_string(), good, quantity).await?;

        db::persist_transaction(pg_pool.clone(), "purchase", user_id, &purchase_order).await?;

        Ok(purchase_order)
    } else {
        Err(anyhow!("Refusing to try and create a purchase order with zero quantity"))
    }
}

pub async fn create_sell_order(client: Client, pg_pool: PgPool, user_id: &str, ship_id: &str, good: Good, quantity: i32) -> anyhow::Result<responses::PurchaseOrder> {
    if quantity > 0 {
        let sell_order = client.create_sell_order(ship_id.to_string(), good, quantity).await?;

        db::persist_transaction(pg_pool.clone(), "sell", user_id, &sell_order).await?;

        Ok(sell_order)
    } else {
        Err(anyhow!("Refusing to try and create a sell order with zero quantity"))
    }
}

pub async fn get_additional_fuel_required_for_trip(pg_pool: PgPool, http_client: Client, ship_id: &str, ship_type: &str, current_fuel: i32, origin: &str, destination: &str) -> anyhow::Result<i32> {
    let db_fuel_required = db::get_fuel_required(pg_pool.clone(), origin, destination, ship_type).await?;

    // if we already have already made this flight before then we know exactly how much fuel is required
    if let Some(db_fuel_required) = db_fuel_required {
        return Ok(db_fuel_required - current_fuel)
    }

    match http_client.create_flight_plan(ship_id.to_string(), destination.to_string()).await {
        Ok(_) => {
            log::error!("This request should have failed. This method should only be called with ships that have zero fuel");

            Err(anyhow!("Request shouldn't have succeeded. Ship is now in motion"))
        },
        Err(e) => match e {
            SpaceTradersClientError::ApiError(error_message) => {
                let re = Regex::new(r"You require (\d+?) more FUEL").unwrap();

                let mut additional_fuel_required = 0;
                for cap in re.captures_iter(&error_message.error.message) {
                    additional_fuel_required = cap[1].parse::<i32>().unwrap();
                }

                Ok(additional_fuel_required)
            },
            e => {
                Err(anyhow!("Expected a \"not enough fuel\" error, but got {} instead", e))
            },
        }
    }

    // TODO: Rather than try and guess the formula lets attempt to create the flight plan
    //       (after ensuring that the ship has no fuel) and then parse the response to get the
    //       actual fuel requirements. After the flight plan is written to the db then we will
    //       just use that value instead of this process again.

    // let distance_between = db::get_distance_between_locations(pg_pool, origin, destination).await?;

    // https://discord.com/channels/792864705139048469/792864705139048472/839919413742272572
    // floor((cargo - fuelRequired) / volume) * (sell - buy) / time
    // https://discord.com/channels/792864705139048469/792864705139048472/836090525307371541
    // time = distance * (2 / speed) + 60
    // let planet_penalty = if distance_between.origin_location_type == "Planet" { 2.0 } else { 0.0 };
    // let fuel_required: f64 = (distance_between.distance.round() / 4.0) + planet_penalty + 1.0;

    // let ship_fuel_penalty = match ship_type {
    //     "GR-MK-II" => 1.0,
    //     "GR-MK-III" => 2.0,
    //     _ => 0.0,
    // };

    // Ok((fuel_required + ship_fuel_penalty).ceil() as i32)
}

pub async fn get_routes_for_ship(pg_pool: PgPool, location_symbol: &str, ship_speed: i32) -> anyhow::Result<Vec<DbRoute>> {
    // TODO: Getting the best route only from the location that the ship currently is in locks
    //       the ship into trade loops. It might be better to search the entire system for the
    //       best route and then find the best trade to that location before beginning a trade
    //       route. That way we move around the system a little more
    match db::get_routes_from_location(pg_pool.clone(), location_symbol, ship_speed).await {
        Ok(routes) => return Ok(routes),
        Err(e) => panic!("Unable to get routes for ship {:?}", e),
    };
}
